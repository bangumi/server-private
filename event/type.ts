export const EventOp = Object.freeze({
  Create: 'c',
  Delete: 'd',
  Update: 'u',
  Snapshot: 'r',
} as const);

export type EventOp = (typeof EventOp)[keyof typeof EventOp];

export interface Payload {
  // before: object;
  // after: object;
  source: {
    table: string;
  };
  op: EventOp;
}

export interface KafkaMessage {
  topic: string;
  key: string;
  value: string;
}
