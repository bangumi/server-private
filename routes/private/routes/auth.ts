import { createError } from '@fastify/error';
import { Type as t } from '@sinclair/typebox';
import httpCodes from 'http-status-codes';

import { comparePassword, NeedLoginError } from '@app/lib/auth/index.ts';
import * as session from '@app/lib/auth/session.ts';
import { CookieKey } from '@app/lib/auth/session.ts';
import config, { redisPrefix } from '@app/lib/config.ts';
import { BadRequestError, CaptchaError } from '@app/lib/error.ts';
import { avatar } from '@app/lib/images';
import { Tag } from '@app/lib/openapi/index.ts';
import { fetchPermission, UserRepo } from '@app/lib/orm/index.ts';
import { createTurnstileDriver } from '@app/lib/services/turnstile.ts';
import * as res from '@app/lib/types/res.ts';
import { createLimiter } from '@app/lib/utils/rate-limit/index.ts';
import { requireLogin } from '@app/routes/hooks/pre-handler.ts';
import type { App } from '@app/routes/type.ts';

const TooManyRequestsError = createError(
  'TOO_MANY_REQUESTS',
  'too many failed login attempts',
  httpCodes.TOO_MANY_REQUESTS,
);

const EmailOrPasswordError = createError(
  'EMAIL_PASSWORD_ERROR',
  'email does not exists or email and password not match',
  httpCodes.UNAUTHORIZED,
);

const UserBannedError = createError('USER_BANNED', 'user is banned', httpCodes.UNAUTHORIZED);

const allowedRedirectUris: string[] = ['bangumi://', 'ani://bangumi-turnstile-callback'];

// eslint-disable-next-line @typescript-eslint/require-await
export async function setup(app: App) {
  // 10 calls per 600s
  const limiter = createLimiter();

  app.post(
    '/logout',
    {
      schema: {
        description: '登出',
        operationId: 'logout',
        tags: [Tag.Auth],
        body: t.Object({}),
        response: {
          200: {},
          401: res.Ref(res.Error, {
            description: '未登录',
            'x-examples': {
              NeedLoginError: { value: res.formatError(new NeedLoginError('logout')) },
            },
          }),
        },
      },
      preHandler: [requireLogin('logout')],
    },
    async (req, res) => {
      if (!req.auth.login) {
        throw new NeedLoginError('logout');
      }

      const sessionKey = req.cookies[session.CookieKey];

      if (!sessionKey) {
        return;
      }

      await session.revoke(sessionKey);
      void res.clearCookie(CookieKey);
    },
  );

  const turnstile = createTurnstileDriver(config.turnstile.secretKey);

  const loginRequestBody = t.Object(
    {
      email: t.String({ minLength: 1 }),
      password: t.String({ minLength: 1 }),
      'cf-turnstile-response': t.String({ minLength: 1 }),
    },
    {
      $id: 'LoginRequestBody',
      examples: [
        {
          email: 'treeholechan@gmail.com',
          password: 'lovemeplease',
          'cf-turnstile-response': '10000000-aaaa-bbbb-cccc-000000000001',
        },
      ],
    },
  );

  app.addSchema(loginRequestBody);

  app.post(
    '/login',
    {
      schema: {
        description: `需要 [turnstile](https://developers.cloudflare.com/turnstile/get-started/client-side-rendering/)

next.bgm.tv 域名对应的 site-key 为 \`0x4AAAAAAABkMYinukE8nzYS\`

dev.bgm38.com 域名使用测试用的 site-key \`1x00000000000000000000AA\``,
        operationId: 'login',
        tags: [Tag.Auth],
        response: {
          200: res.Ref(res.SlimUser, {
            headers: {
              'Set-Cookie': t.String({ description: `example: "${session.CookieKey}=12345abc"` }),
            },
          }),
          400: res.Ref(res.Error, { description: 'request validation error' }),
          401: res.Ref(res.Error, {
            description: '验证码错误/账号密码不匹配',
            headers: {
              'X-RateLimit-Remaining': t.Integer({ description: 'remaining rate limit' }),
              'X-RateLimit-Limit': t.Integer({ description: 'total limit per 10 minutes' }),
              'X-RateLimit-Reset': t.Integer({ description: 'seconds to reset rate limit' }),
            },
            'x-examples': res.formatErrors(new CaptchaError(), new EmailOrPasswordError()),
          }),
          429: res.Ref(res.Error, {
            description: '失败次数太多，需要过一段时间再重试',
            headers: {
              'X-RateLimit-Remaining': t.Integer({ description: 'remaining rate limit' }),
              'X-RateLimit-Limit': t.Integer({ description: 'limit per 10 minutes' }),
              'X-RateLimit-Reset': t.Integer({ description: 'seconds to reset rate limit' }),
            },
            examples: [res.formatError(new TooManyRequestsError())],
          }),
        },
        body: res.Ref(loginRequestBody),
      },
    },
    async function loginHandler(
      { body: { email, password, 'cf-turnstile-response': cfCaptchaResponse }, ip },
      reply,
    ): Promise<res.ISlimUser> {
      const limitKey = `${redisPrefix}-login-rate-limit-${ip}`;
      const { remain, reset, limit, limited } = await limiter.get(limitKey, 600, 10);
      void reply.headers({
        'X-RateLimit-Remaining': remain,
        'X-RateLimit-Limit': limit,
        'X-RateLimit-Reset': reset,
      });
      if (limited) {
        throw new TooManyRequestsError();
      }

      if (!(await turnstile.verify(cfCaptchaResponse))) {
        throw new CaptchaError();
      }

      const user = await UserRepo.findOne({ where: { email } });
      if (!user) {
        throw new EmailOrPasswordError();
      }
      if (!(await comparePassword(user.passwordCrypt, password))) {
        throw new EmailOrPasswordError();
      }

      const perms = await fetchPermission(user.groupid);
      if (perms.user_ban) {
        throw new UserBannedError();
      }

      const token = await session.create({
        id: user.id,
        regTime: user.regdate,
      });

      await limiter.reset(limitKey);

      void reply.cookie(CookieKey, token, { maxAge: 24 * 60 * 60 * 30 });

      return {
        ...user,
        avatar: avatar(user.avatar),
        joinedAt: user.regdate,
      };
    },
  );

  app.get(
    '/turnstile',
    {
      schema: {
        summary: '获取 Turnstile 令牌',
        description: '为防止滥用，Redirect URI 为白名单机制，如需添加请提交 PR。',
        operationId: 'getTurnstileToken',
        tags: [Tag.Auth],
        querystring: t.Object({
          theme: t.Optional(
            t.Enum({
              dark: 'dark',
              light: 'light',
              auto: 'auto',
            }),
          ),
          redirect_uri: t.String(),
        }),
      },
    },
    async (req, res) => {
      const redirectUri = req.query.redirect_uri;
      try {
        new URL(redirectUri);
      } catch {
        throw BadRequestError('Invalid redirect URI.');
      }
      if (!allowedRedirectUris.some((allowedUri) => redirectUri.startsWith(allowedUri))) {
        throw BadRequestError(
          `Redirect URI is not in the whitelist, you can PR your redirect URI.`,
        );
      }
      await res.view('turnstile', {
        TURNSTILE_SITE_KEY: config.turnstile.siteKey,
        turnstile_theme: req.query.theme || 'auto',
        redirect_uri: Buffer.from(redirectUri).toString('base64'),
      });
    },
  );
}
