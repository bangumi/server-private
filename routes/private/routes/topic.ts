import type { Static } from '@sinclair/typebox';
import { Type as t } from '@sinclair/typebox';

import type { IAuth } from '@app/lib/auth';
import { NotAllowedError } from '@app/lib/auth';
import { Dam, dam } from '@app/lib/dam';
import { BadRequestError, NotFoundError, UnexpectedNotFoundError } from '@app/lib/error';
import * as Like from '@app/lib/like';
import { Security, Tag } from '@app/lib/openapi';
import type { Page } from '@app/lib/orm';
import * as orm from '@app/lib/orm';
import { GroupMemberRepo, isMemberInGroup } from '@app/lib/orm';
import { avatar, groupIcon } from '@app/lib/response';
import type { ITopic } from '@app/lib/topic';
import * as Topic from '@app/lib/topic';
import { CommentState, NotJoinPrivateGroupError, TopicDisplay } from '@app/lib/topic';
import * as res from '@app/lib/types/res';
import { formatErrors, toResUser } from '@app/lib/types/res';
import { requireLogin } from '@app/routes/hooks/pre-handler';
import type { App } from '@app/routes/type';

const Group = t.Object(
  {
    id: t.Integer(),
    name: t.String(),
    nsfw: t.Boolean(),
    title: t.String(),
    icon: t.String(),
    description: t.String(),
    totalMembers: t.Integer(),
    createdAt: t.Integer(),
  },
  { $id: 'Group' },
);

type IGroupMember = Static<typeof GroupMember>;
const GroupMember = t.Object(
  {
    avatar: res.Avatar,
    id: t.Integer(),
    nickname: t.String(),
    username: t.String(),
    joinedAt: t.Integer(),
  },
  { $id: 'GroupMember' },
);
const SubReply = t.Object(
  {
    id: t.Integer(),
    creator: t.Ref(res.User),
    createdAt: t.Integer(),
    isFriend: t.Boolean(),
    text: t.String(),
    state: t.Integer(),
  },
  { $id: 'SubReply' },
);

const Reaction = t.Object({
  selected: t.Boolean(),
  total: t.Integer(),
  value: t.Integer(),
});

const Reply = t.Object(
  {
    id: t.Integer(),
    isFriend: t.Boolean(),
    replies: t.Array(t.Ref(SubReply)),
    creator: t.Ref(res.User),
    createdAt: t.Integer(),
    text: t.String(),
    state: t.Integer(),
    reactions: t.Array(Reaction),
  },
  { $id: 'Reply' },
);

const TopicDetail = t.Object(
  {
    id: t.Integer(),
    group: t.Ref(Group),
    creator: t.Ref(res.User),
    title: t.String(),
    text: t.String(),
    state: t.Integer(),
    createdAt: t.Integer(),
    replies: t.Array(t.Ref(Reply)),
    reactions: t.Array(Reaction),
  },
  { $id: 'TopicDetail' },
);

const TopicBasic = t.Object(
  {
    title: t.String({ minLength: 1 }),
    text: t.String({ minLength: 1, description: 'bbcode' }),
  },
  { $id: 'TopicCreation', examples: [{ title: 'topic title', content: 'topic content' }] },
);

// eslint-disable-next-line @typescript-eslint/require-await
export async function setup(app: App) {
  app.addSchema(res.Error);
  app.addSchema(res.Topic);
  app.addSchema(Group);
  app.addSchema(TopicBasic);

  const GroupProfile = t.Object(
    {
      recentAddedMembers: t.Array(t.Ref(GroupMember)),
      topics: t.Array(t.Ref(res.Topic)),
      inGroup: t.Boolean({ description: '是否已经加入小组' }),
      group: t.Ref(Group),
      totalTopics: t.Integer(),
    },
    { $id: 'GroupProfile' },
  );

  app.addSchema(GroupProfile);

  app.get(
    '/groups/:groupName/profile',
    {
      schema: {
        description: '获取小组首页',
        operationId: 'getGroupProfile',
        tags: [Tag.Topic],
        params: t.Object({
          groupName: t.String({ minLength: 1 }),
        }),
        querystring: t.Object({
          limit: t.Optional(t.Integer({ default: 20, maximum: 40 })),
          offset: t.Optional(t.Integer({ default: 0, minimum: 0 })),
        }),
        response: {
          200: t.Ref(GroupProfile),
          404: t.Ref(res.Error, {
            description: '小组不存在',
            'x-examples': {
              NotFoundError: { value: res.formatError(NotFoundError('topic')) },
            },
          }),
        },
      },
    },
    async ({ params, auth, query }) => {
      const group = await orm.fetchGroup(params.groupName);

      if (!group) {
        throw new NotFoundError('group');
      }

      const [total, topicList] = await Topic.fetchTopicList(auth, 'group', group.id, query);

      const topics = await addCreators(topicList, group.id);

      return {
        group: { ...group, icon: groupIcon(group.icon) },
        totalTopics: total,
        inGroup: auth.login ? await orm.isMemberInGroup(group.id, auth.userID) : false,
        topics,
        recentAddedMembers: await fetchRecentMember(group.id),
      };
    },
  );

  app.addSchema(SubReply);

  app.addSchema(Reply);

  app.addSchema(TopicDetail);

  app.get(
    '/groups/-/topics/:id',
    {
      schema: {
        description: '获取帖子列表',
        operationId: 'getGroupTopicDetail',
        tags: [Tag.Topic],
        params: t.Object({
          id: t.Integer({ examples: [371602] }),
        }),
        response: {
          200: t.Ref(TopicDetail),
          404: t.Ref(res.Error, {
            description: '小组不存在',
            'x-examples': {
              NotFoundError: { value: res.formatError(NotFoundError('topic')) },
            },
          }),
        },
      },
    },
    handleTopicDetail,
  );

  app.addSchema(GroupMember);

  app.get(
    '/groups/:groupName/members',
    {
      schema: {
        description: '获取帖子列表',
        operationId: 'listGroupMembersByName',
        tags: [Tag.Topic],
        params: t.Object({
          groupName: t.String({ minLength: 1 }),
        }),
        querystring: t.Object({
          type: t.Optional(
            t.Enum(
              {
                mod: 'mod',
                normal: 'normal',
                all: 'all',
              } as const,
              { default: 'all' },
            ),
          ),
          limit: t.Optional(t.Integer({ default: 30, maximum: 40 })),
          offset: t.Optional(t.Integer({ default: 0 })),
        }),
        response: {
          200: res.Paged(t.Ref(GroupMember)),
          404: t.Ref(res.Error, {
            description: '小组不存在',
            'x-examples': {
              NotFoundError: { value: res.formatError(new NotFoundError('topic')) },
            },
          }),
        },
      },
    },
    async ({ params, query: { type = 'all', limit, offset } }) => {
      const group = await orm.fetchGroup(params.groupName);

      if (!group) {
        throw new NotFoundError('group');
      }

      const [total, members] = await fetchGroupMemberList(group.id, { type, limit, offset });

      return { total, data: members };
    },
  );

  app.get(
    '/groups/:groupName/topics',
    {
      schema: {
        description: '获取帖子列表',
        operationId: 'getGroupTopicsByGroupName',
        tags: [Tag.Topic],
        params: t.Object({
          groupName: t.String({ minLength: 1 }),
        }),
        querystring: t.Object({
          limit: t.Optional(t.Integer({ default: 30, maximum: 40 })),
          offset: t.Optional(t.Integer({ default: 0 })),
        }),
        response: {
          200: res.Paged(t.Ref(res.Topic)),
          404: t.Ref(res.Error, {
            description: '小组不存在',
            'x-examples': {
              NotFoundError: { value: res.formatError(NotFoundError('topic')) },
            },
          }),
        },
      },
    },
    async ({ params, query, auth }) => {
      const group = await orm.fetchGroup(params.groupName);

      if (!group) {
        throw new NotFoundError('group');
      }

      if (!group.accessible && !(await isMemberInGroup(group.id, auth.userID))) {
        throw new NotJoinPrivateGroupError(group.name);
      }

      const [total, topics] = await Topic.fetchTopicList(auth, 'group', group.id, query);

      return { total, data: await addCreators(topics, group.id) };
    },
  );

  app.get(
    '/subjects/:subjectID/topics',
    {
      schema: {
        description: '获取帖子列表',
        operationId: 'getSubjectTopicsBySubjectId',
        tags: [Tag.Topic],
        params: t.Object({
          subjectID: t.Integer({ exclusiveMinimum: 0 }),
        }),
        querystring: t.Object({
          limit: t.Optional(t.Integer({ default: 30, maximum: 40 })),
          offset: t.Optional(t.Integer({ default: 0 })),
        }),
        response: {
          200: res.Paged(t.Ref(res.Topic)),
          404: t.Ref(res.Error, {
            description: '条目不存在',
            'x-examples': {
              NotFoundError: { value: res.formatError(NotFoundError('topic')) },
            },
          }),
        },
      },
    },
    async ({ params: { subjectID }, query, auth }) => {
      const subject = await orm.fetchSubject(subjectID);
      if (!subject) {
        throw new NotFoundError(`subject ${subjectID}`);
      }
      if (subject.nsfw && !auth.allowNsfw) {
        throw new NotFoundError(`subject ${subjectID}`);
      }

      const [total, topics] = await Topic.fetchTopicList(auth, 'subject', subjectID, query);
      return { total, data: await addCreators(topics, subjectID) };
    },
  );

  app.post(
    '/groups/:groupName/topics',
    {
      schema: {
        tags: [Tag.Group],
        operationId: 'createNewGroupTopic',
        params: t.Object({
          groupName: t.String({ minLength: 1, examples: ['sandbox'] }),
        }),
        response: {
          200: t.Object({
            id: t.Integer({ description: 'new topic id' }),
          }),
        },
        security: [{ [Security.CookiesSession]: [] }],
        body: t.Ref(TopicBasic),
      },
      preHandler: [requireLogin('creating a post')],
    },
    async ({ auth, body: { text, title }, params: { groupName } }) => {
      if (auth.permission.ban_post) {
        throw new NotAllowedError('create posts');
      }

      const group = await orm.fetchGroup(groupName);
      if (!group) {
        throw new NotFoundError(`group ${groupName}`);
      }

      let display = TopicDisplay.Normal;

      if (dam.needReview(title) || dam.needReview(text)) {
        display = TopicDisplay.Review;
      }

      if (!group.accessible && !(await orm.isMemberInGroup(group.id, auth.userID))) {
        throw new NotAllowedError('create posts, join group first');
      }

      return await orm.createPostInGroup({
        title,
        content: text,
        display,
        userID: auth.userID,
        groupID: group.id,
        state: Topic.CommentState.Normal,
      });
    },
  );

  app.put(
    '/groups/-/topics/:topicID',
    {
      schema: {
        operationId: 'editGroupTopic',
        params: t.Object({
          topicID: t.Integer({ examples: [371602] }),
        }),
        tags: [Tag.Group],
        response: {
          200: t.Object({}),
          400: t.Ref(res.Error),
          401: t.Ref(res.Error, {
            'x-examples': formatErrors(NotAllowedError('edit a topic')),
          }),
        },
        security: [{ [Security.CookiesSession]: [] }],
        body: t.Ref(TopicBasic),
      },
      preHandler: [requireLogin('edit a topic')],
    },
    /**
     * @param auth -
     * @param title - 帖子标题
     * @param text - 帖子内容
     * @param topicID - 帖子 ID
     */
    async function ({
      auth,
      body: { title, text },
      params: { topicID },
    }): Promise<Record<string, never>> {
      if (auth.permission.ban_post) {
        throw new NotAllowedError('create reply');
      }

      if (!(Dam.allCharacterPrintable(title) && Dam.allCharacterPrintable(text))) {
        throw new BadRequestError('text contains invalid invisible character');
      }

      const topic = await Topic.fetchDetail(auth, 'group', topicID);
      if (!topic) {
        throw new NotFoundError(`topic ${topicID}`);
      }

      if (
        ![CommentState.AdminReopen, CommentState.AdminPin, CommentState.Normal].includes(
          topic.state,
        )
      ) {
        throw new NotAllowedError('edit this topic');
      }

      if (topic.creatorID !== auth.userID) {
        throw new NotAllowedError('edit this topic');
      }

      let display = topic.display;
      if (dam.needReview(title) || dam.needReview(text)) {
        if (display === TopicDisplay.Normal) {
          display = TopicDisplay.Review;
        } else {
          return {};
        }
      }

      await orm.GroupTopicRepo.update({ id: topicID }, { title, display });

      const topicPost = await orm.GroupPostRepo.findOneBy({ topicID });

      if (topicPost) {
        await orm.GroupPostRepo.update({ id: topicPost.id }, { content: text });
      }

      return {};
    },
  );
}

async function addCreators(
  topics: ITopic[],
  parentID: number,
): Promise<Static<typeof res.Topic>[]> {
  const withCreator = await orm.addCreator(topics);

  return withCreator.map((x) => {
    return {
      ...x,
      creator: toResUser(x.creator),
      updatedAt: x.updatedAt,
      parentID,
    };
  });
}

async function fetchGroupMemberList(
  groupID: number,
  { limit = 30, offset = 0, type }: Page & { type: 'mod' | 'normal' | 'all' },
): Promise<[number, IGroupMember[]]> {
  const where = {
    gmbGid: groupID,
    gmbModerator: type === 'all' ? undefined : type === 'mod',
  } as const;
  const total = await GroupMemberRepo.count({ where });

  const members = await GroupMemberRepo.find({
    where,
    take: limit,
    skip: offset,
    order: {
      gmbDateline: 'desc',
    },
  });

  const users = await orm.fetchUsers(members.map((x) => x.gmbUid));

  return [
    total,
    members.map(function (x): IGroupMember {
      const user = users[x.gmbUid];
      if (!user) {
        throw new UnexpectedNotFoundError(`user ${x.gmbUid}`);
      }

      return {
        avatar: avatar(user.img),
        id: user.id,
        joinedAt: x.gmbDateline,
        nickname: user.nickname,
        username: user.username,
      };
    }),
  ];
}

async function fetchRecentMember(groupID: number): Promise<IGroupMember[]> {
  const [_, members] = await fetchGroupMemberList(groupID, { limit: 6, type: 'all' });

  return members;
}

export async function handleTopicDetail({
  params: { id },
  auth,
}: {
  params: { id: number };
  auth: IAuth;
}): Promise<Static<typeof TopicDetail>> {
  const topic = await Topic.fetchDetail(auth, 'group', id);
  if (!topic) {
    throw new NotFoundError(`topic ${id}`);
  }

  const group = await orm.fetchGroupByID(topic.parentID);
  if (!group) {
    throw new UnexpectedNotFoundError(`group ${topic.parentID}`);
  }

  const userIds: number[] = [
    topic.creatorID,
    ...topic.replies.flatMap((x) => [x.creatorID, ...x.replies.map((x) => x.creatorID)]),
  ];

  const friends = await orm.fetchFriends(auth.userID);
  const users = await orm.fetchUsers([...new Set(userIds)]);

  const creator = users[topic.creatorID];
  if (!creator) {
    throw new UnexpectedNotFoundError(`user ${topic.creatorID}`);
  }

  const reactions = await Like.fetchGroupTopic(id, auth.userID);

  return {
    ...topic,
    creator: toResUser(creator),
    text: topic.text,
    group: { ...group, icon: groupIcon(group.icon) },
    reactions: reactions[topic.contentPost.id] ?? [],
    replies: topic.replies.map((x) => {
      const user = users[x.creatorID];
      if (!user) {
        throw new UnexpectedNotFoundError(`user ${x.creatorID}`);
      }
      return {
        reactions: reactions[x.id] ?? [],
        isFriend: friends[x.creatorID] ?? false,
        ...x,
        replies: x.replies.map((x) => {
          const user = users[x.creatorID];
          if (!user) {
            throw new UnexpectedNotFoundError(`user ${x.creatorID}`);
          }
          return {
            reactions: reactions[x.id] ?? [],
            isFriend: friends[x.creatorID] ?? false,
            ...x,
            creator: toResUser(user),
          };
        }),
        creator: {
          isFriend: friends[x.creatorID] ?? false,
          ...toResUser(user),
        },
      };
    }),
    state: topic.state,
  };
}
