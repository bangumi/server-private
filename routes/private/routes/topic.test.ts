import { fastify } from 'fastify';
import { describe, expect, test, vi } from 'vitest';

import type { IAuth } from '@app/lib/auth';
import { emptyAuth, UserGroup } from '@app/lib/auth';
import * as orm from '@app/lib/orm';
import { createServer } from '@app/lib/server';
import { fetchDetail } from '@app/lib/topic';
import { createTestServer } from '@app/tests/utils';

import { setup } from './topic';

const expectedTopic = {
  createdAt: 1657885648,
  creator: {
    avatar: {
      large: 'https://lain.bgm.tv/pic/user/l/icon.jpg',
      medium: 'https://lain.bgm.tv/pic/user/m/icon.jpg',
      small: 'https://lain.bgm.tv/pic/user/s/icon.jpg',
    },
    id: 287622,
    nickname: 'nickname 287622',
    sign: 'sing 287622',
    user_group: 0,
    username: '287622',
  },
  id: 371602,
  parentID: 4215,
  repliesCount: 2,
  title: 'tes',
  updatedAt: 1662283112,
};

describe('group topics', () => {
  test('should failed on not found group', async () => {
    const app = await createServer();

    const res = await app.inject({
      url: '/p1/groups/non-existing/topics',
    });

    expect(res.statusCode).toBe(404);
    expect(res.json()).toMatchSnapshot();
  });

  test('should return data', async () => {
    const app = await createServer();

    const res = await app.inject({
      url: '/p1/groups/sandbox/topics',
    });
    const data = res.json();

    expect(res.statusCode).toBe(200);
    expect(data.data).toContainEqual(expectedTopic);
  });

  test('should fetch group profile', async () => {
    const app = await createServer();

    const res = await app.inject('/p1/groups/sandbox/profile');
    const data = res.json();

    expect(res.statusCode).toBe(200);
    expect(data.group).toEqual({
      createdAt: 1531631310,
      description: '[s]非[/s]官方沙盒',
      icon: 'https://lain.bgm.tv/pic/icon/s/000/00/42/4215.jpg?r=1531631345',
      id: 4215,
      name: 'sandbox',
      nsfw: false,
      title: '沙盒',
      totalMembers: 3,
    });
    expect(data.topics).toContainEqual(expectedTopic);
  });

  test('should fetch topic details', async () => {
    const app = await createServer();

    const res = await app.inject('/p1/groups/-/topics/371602');
    expect(res.statusCode).toBe(200);
    expect(res.json()).toMatchSnapshot();
  });
});

describe('create group post', () => {
  const createPostInGroup = vi.fn().mockResolvedValue({ id: 1 });
  vi.spyOn(orm, 'createPostInGroup').mockImplementation(createPostInGroup);

  test('should create group topic', async () => {
    const app = fastify();

    app.addHook('preHandler', (req, res, done) => {
      req.auth = {
        groupID: UserGroup.Normal,
        login: true,
        permission: {},
        allowNsfw: true,
        regTime: 0,
        userID: 100,
      } satisfies IAuth;
      done();
    });

    await app.register(setup);

    const res = await app.inject({
      url: '/groups/sandbox/topics',
      method: 'post',
      payload: {
        title: 'post title',
        text: 'post contents',
      },
    });

    expect(res.json()).toMatchObject({ id: 1 });
    expect(res.statusCode).toBe(200);
  });

  test('should not create with banned user', async () => {
    const app = fastify();

    app.addHook('preHandler', (req, res, done) => {
      req.auth = {
        groupID: UserGroup.Normal,
        login: true,
        permission: {
          ban_post: true,
        },
        allowNsfw: true,
        regTime: 0,
        userID: 1,
      } satisfies IAuth;
      done();
    });

    await app.register(setup);

    const res = await app.inject({
      url: '/groups/sandbox/topics',
      method: 'post',
      payload: {
        title: 'post title',
        text: 'post contents',
      },
    });

    expect(res.statusCode).toBe(401);
    expect(createPostInGroup).toBeCalledTimes(1);
  });
});

describe('edit topic', () => {
  test('should edit topic', async () => {
    const app = createTestServer({
      auth: {
        ...emptyAuth(),
        login: true,
        userID: 287622,
      },
    });

    await app.register(setup);

    {
      const res = await app.inject({
        url: '/groups/-/topics/375793',
        method: 'put',
        payload: {
          title: 'new topic title',
          text: 'new contents',
        },
      });

      expect(res.statusCode).toBe(200);

      const topic = await fetchDetail(emptyAuth(), 'group', 375793);

      expect(topic?.title).toBe('new topic title');
      expect(topic?.text).toBe('new contents');
    }

    {
      const res = await app.inject({
        url: '/groups/-/topics/375793',
        method: 'put',
        payload: {
          title: 'new topic title 2',
          text: 'new contents 2',
        },
      });

      expect(res.statusCode).toBe(200);

      const topic = await fetchDetail(emptyAuth(), 'group', 375793);

      expect(topic?.title).toBe('new topic title 2');
      expect(topic?.text).toBe('new contents 2');
    }
  });

  test('should not edited topic by non-owner', async () => {
    const app = createTestServer({
      auth: {
        ...emptyAuth(),
        login: true,
        userID: 1,
      },
    });

    await app.register(setup);

    const res = await app.inject({
      url: '/groups/-/topics/371602',
      method: 'put',
      payload: {
        title: 'new topic title',
        text: 'new contents',
      },
    });

    expect(res.json()).toMatchInlineSnapshot(`
      Object {
        "code": "NOT_ALLOWED",
        "error": "Unauthorized",
        "message": "you don't have permission to edit this topic",
        "statusCode": 401,
      }
    `);
    expect(res.statusCode).toBe(401);
  });
});
