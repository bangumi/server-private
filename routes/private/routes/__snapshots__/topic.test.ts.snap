// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`group topics > should failed on not found group 1`] = `
Object {
  "code": "NOT_FOUND",
  "error": "Not Found",
  "message": "group not found",
  "statusCode": 404,
}
`;

exports[`group topics > should fetch topic details 1`] = `
Object {
  "createdAt": 1657885648,
  "creator": Object {
    "avatar": Object {
      "large": "https://lain.bgm.tv/pic/user/l/icon.jpg",
      "medium": "https://lain.bgm.tv/pic/user/m/icon.jpg",
      "small": "https://lain.bgm.tv/pic/user/s/icon.jpg",
    },
    "id": 287622,
    "nickname": "nickname 287622",
    "sign": "sing 287622",
    "user_group": 0,
    "username": "287622",
  },
  "group": Object {
    "createdAt": 1531631310,
    "description": "[s]非[/s]官方沙盒",
    "icon": "https://lain.bgm.tv/pic/icon/s/000/00/42/4215.jpg?r=1531631345",
    "id": 4215,
    "name": "sandbox",
    "nsfw": false,
    "title": "沙盒",
    "totalMembers": 3,
  },
  "id": 371602,
  "reactions": Array [],
  "replies": Array [
    Object {
      "createdAt": 1658030918,
      "creator": Object {
        "avatar": Object {
          "large": "https://lain.bgm.tv/pic/user/l/icon.jpg",
          "medium": "https://lain.bgm.tv/pic/user/m/icon.jpg",
          "small": "https://lain.bgm.tv/pic/user/s/icon.jpg",
        },
        "id": 287622,
        "nickname": "nickname 287622",
        "sign": "sing 287622",
        "user_group": 0,
        "username": "287622",
      },
      "id": 2056216,
      "isFriend": false,
      "reactions": Array [],
      "replies": Array [],
      "state": 6,
      "text": "",
    },
    Object {
      "createdAt": 1662276469,
      "creator": Object {
        "avatar": Object {
          "large": "https://lain.bgm.tv/pic/user/l/icon.jpg",
          "medium": "https://lain.bgm.tv/pic/user/m/icon.jpg",
          "small": "https://lain.bgm.tv/pic/user/s/icon.jpg",
        },
        "id": 287622,
        "nickname": "nickname 287622",
        "sign": "sing 287622",
        "user_group": 0,
        "username": "287622",
      },
      "id": 2092004,
      "isFriend": false,
      "reactions": Array [],
      "replies": Array [],
      "state": 7,
      "text": "",
    },
    Object {
      "createdAt": 1662282514,
      "creator": Object {
        "avatar": Object {
          "large": "https://lain.bgm.tv/pic/user/l/icon.jpg",
          "medium": "https://lain.bgm.tv/pic/user/m/icon.jpg",
          "small": "https://lain.bgm.tv/pic/user/s/icon.jpg",
        },
        "id": 1,
        "nickname": "nickname 1",
        "sign": "sing 1",
        "user_group": 0,
        "username": "1",
      },
      "id": 2092065,
      "isFriend": false,
      "reactions": Array [],
      "replies": Array [],
      "state": 5,
      "text": "",
    },
    Object {
      "createdAt": 1662282518,
      "creator": Object {
        "avatar": Object {
          "large": "https://lain.bgm.tv/pic/user/l/icon.jpg",
          "medium": "https://lain.bgm.tv/pic/user/m/icon.jpg",
          "small": "https://lain.bgm.tv/pic/user/s/icon.jpg",
        },
        "id": 1,
        "nickname": "nickname 1",
        "sign": "sing 1",
        "user_group": 0,
        "username": "1",
      },
      "id": 2092066,
      "isFriend": false,
      "reactions": Array [],
      "replies": Array [],
      "state": 1,
      "text": "",
    },
    Object {
      "createdAt": 1662282523,
      "creator": Object {
        "avatar": Object {
          "large": "https://lain.bgm.tv/pic/user/l/icon.jpg",
          "medium": "https://lain.bgm.tv/pic/user/m/icon.jpg",
          "small": "https://lain.bgm.tv/pic/user/s/icon.jpg",
        },
        "id": 1,
        "nickname": "nickname 1",
        "sign": "sing 1",
        "user_group": 0,
        "username": "1",
      },
      "id": 2092067,
      "isFriend": false,
      "reactions": Array [],
      "replies": Array [],
      "state": 2,
      "text": "",
    },
    Object {
      "createdAt": 1662282554,
      "creator": Object {
        "avatar": Object {
          "large": "https://lain.bgm.tv/pic/user/l/icon.jpg",
          "medium": "https://lain.bgm.tv/pic/user/m/icon.jpg",
          "small": "https://lain.bgm.tv/pic/user/s/icon.jpg",
        },
        "id": 287622,
        "nickname": "nickname 287622",
        "sign": "sing 287622",
        "user_group": 0,
        "username": "287622",
      },
      "id": 2092069,
      "isFriend": false,
      "reactions": Array [],
      "replies": Array [
        Object {
          "createdAt": 1662283112,
          "creator": Object {
            "avatar": Object {
              "large": "https://lain.bgm.tv/pic/user/l/icon.jpg",
              "medium": "https://lain.bgm.tv/pic/user/m/icon.jpg",
              "small": "https://lain.bgm.tv/pic/user/s/icon.jpg",
            },
            "id": 287622,
            "nickname": "nickname 287622",
            "sign": "sing 287622",
            "user_group": 0,
            "username": "287622",
          },
          "id": 2092074,
          "isFriend": false,
          "reactions": Array [],
          "state": 0,
          "text": "sub",
        },
      ],
      "state": 0,
      "text": "q",
    },
  ],
  "state": 2,
  "text": "tt",
  "title": "tes",
}
`;
